fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/metadata/Index.java

```diff
@@ -183,11 +184,19 @@ public Index(@Nonnull RecordMetaDataProto.Index proto) throws KeyExpression.Dese
 
         if (proto.hasValueExpression()) {
             KeyExpression value = KeyExpression.fromProto(proto.getValueExpression());
-            rootExpression = toKeyWithValueExpression(expr, value);
-        } else {
-            rootExpression = expr;
+            expr = toKeyWithValueExpression(expr, value);
+        }
+
+        final List<KeyExpressionVisitor> visitors = KeyExpressionVisitor.Registry.instance().visitors();
+        if (!visitors.isEmpty()) {
+            for (KeyExpressionVisitor visitor : visitors) {
+                if (visitor.applies(type)) {
+                    expr = visitor.visit(expr);
+                }
+            }
```

Comment:
An alternative to changing the serialized representation to recover information is to use `FunctionKeyExpressions`, which while opaque to most uses could be understood to have that additional Lucene information by the Lucene planner. A `TODO` for now is probably okay, as there is other work in this area outstanding.